File: plugin/src/main/java/net/onelitefeather/pandorascluster/service/PaperCommandService.java
package net.onelitefeather.pandorascluster.service;

import net.kyori.adventure.platform.bukkit.BukkitAudiences;
import net.kyori.adventure.text.format.NamedTextColor;
import net.onelitefeather.pandorascluster.PandorasClusterPlugin;
import net.onelitefeather.pandorascluster.command.LandCommand;
import net.onelitefeather.pandorascluster.command.mapper.PaperSenderMapper;
import org.bukkit.command.CommandSender;
import org.incendo.cloud.annotations.AnnotationParser;
import org.incendo.cloud.execution.ExecutionCoordinator;
import org.incendo.cloud.minecraft.extras.MinecraftHelp;
import org.incendo.cloud.paper.PaperCommandManager;
import org.jetbrains.annotations.NotNull;

public class PaperCommandService {

    private final PandorasClusterPlugin plugin;
    private final PaperCommandManager<CommandSender> paperCommandManager;
    private final AnnotationParser<CommandSender> annotationParser;
    private final MinecraftHelp<CommandSender> minecraftHelp;
    private final BukkitAudiences bukkitAudiences;

    public PaperCommandService(@NotNull PandorasClusterPlugin plugin) {
        this.plugin = plugin;
        this.bukkitAudiences = BukkitAudiences.create(plugin);
        this.paperCommandManager = buildCommandSystem();
        this.annotationParser = buildAnnotationParser();
        this.minecraftHelp = buildHelpSystem();
    }

    @NotNull
    public PaperCommandManager<CommandSender> getPaperCommandManager() {
        return paperCommandManager;
    }

    @NotNull
    public AnnotationParser<CommandSender> getAnnotationParser() {
        return annotationParser;
    }

    @NotNull
    public MinecraftHelp<CommandSender> getMinecraftHelp() {
        return minecraftHelp;
    }

    public void registerCommands() {
        this.annotationParser.parse(new LandCommand(this.plugin.getPandorasCluster()));
    }

    @NotNull
    private MinecraftHelp<CommandSender> buildHelpSystem() {
        return MinecraftHelp.<CommandSender>builder()
                .commandManager(this.paperCommandManager)
                .audienceProvider(this.bukkitAudiences::sender)
                .commandPrefix("/land help")
                .colors(MinecraftHelp.helpColors(
                        NamedTextColor.YELLOW,
                        NamedTextColor.GOLD,
                        NamedTextColor.YELLOW,
                        NamedTextColor.GRAY,
                        NamedTextColor.GOLD))
                .build();
    }

    @NotNull
    private AnnotationParser<CommandSender> buildAnnotationParser() {
        return new AnnotationParser<>(this.paperCommandManager, CommandSender.class);
    }

    @NotNull
    private PaperCommandManager<CommandSender> buildCommandSystem() {
        return PaperCommandManager.builder(new PaperSenderMapper())
                .executionCoordinator(ExecutionCoordinator.asyncCoordinator())
                .buildOnEnable(this.plugin);
    }
}

File: plugin/src/main/kotlin/net/onelitefeather/pandorascluster/service/PaperCommandService.kt
package net.onelitefeather.pandorascluster.service

import cloud.commandframework.annotations.AnnotationParser
import cloud.commandframework.arguments.parser.ParserParameters
import cloud.commandframework.arguments.parser.StandardParameters
import cloud.commandframework.bukkit.CloudBukkitCapabilities
import cloud.commandframework.execution.CommandExecutionCoordinator
import cloud.commandframework.execution.postprocessor.CommandPostprocessingContext
import cloud.commandframework.extra.confirmation.CommandConfirmationManager
import cloud.commandframework.meta.CommandMeta
import cloud.commandframework.minecraft.extras.MinecraftHelp
import cloud.commandframework.paper.PaperCommandManager
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.format.NamedTextColor
import net.onelitefeather.pandorascluster.PandorasClusterPlugin
import net.onelitefeather.pandorascluster.command.commands.*
import net.onelitefeather.pandorascluster.command.parser.LandFlagParser
import net.onelitefeather.pandorascluster.command.parser.LandPlayerParser
import net.onelitefeather.pandorascluster.util.PLUGIN_PREFIX
import org.bukkit.command.CommandSender
import java.util.concurrent.TimeUnit
import java.util.function.Function

class PaperCommandService(private val plugin: PandorasClusterPlugin) {

    lateinit var paperCommandManager: PaperCommandManager<CommandSender>
    lateinit var annotationParser: AnnotationParser<CommandSender>
    lateinit var minecraftHelp: MinecraftHelp<CommandSender>
    lateinit var confirmationManager: CommandConfirmationManager<CommandSender>

    fun setup() {
        paperCommandManager = buildCommandManager()
        annotationParser = buildAnnotationParser()
        confirmationManager = buildConfirmnationManager()
        minecraftHelp = buildHelpSystem()
        registerCommands()
    }

    fun buildCommandManager(): PaperCommandManager<CommandSender> {

        val commandManager = PaperCommandManager.createNative(plugin, CommandExecutionCoordinator.simpleCoordinator())
        if (commandManager.hasCapability(CloudBukkitCapabilities.BRIGADIER)) {
            commandManager.registerBrigadier()
            plugin.logger.info("Brigadier support enabled")
        }

        if (commandManager.hasCapability(CloudBukkitCapabilities.ASYNCHRONOUS_COMPLETION)) {
            commandManager.registerAsynchronousCompletions()
            plugin.logger.info("Asynchronous completions enabled")
        }

        return commandManager
    }

    private fun buildAnnotationParser(): AnnotationParser<CommandSender> {
        val commandMetaFunction = Function<ParserParameters, CommandMeta> { p: ParserParameters ->
            CommandMeta.simple().with(
                CommandMeta.DESCRIPTION,
                p.get(StandardParameters.DESCRIPTION, "No description")
            ).build()
        }
        return AnnotationParser(paperCommandManager, CommandSender::class.java, commandMetaFunction)
    }

    private fun buildConfirmnationManager(): CommandConfirmationManager<CommandSender> {
        val commandConfirmationManager = CommandConfirmationManager(
            20L, TimeUnit.SECONDS, { context: CommandPostprocessingContext<CommandSender> ->
                plugin.bukkitAudiences.sender(context.commandContext.sender).sendMessage(
                    Component.translatable("command.confirm").arguments(PLUGIN_PREFIX)
                )
            },
            { sender: CommandSender ->
                plugin.bukkitAudiences.sender(sender).sendMessage(
                    Component.translatable("command.confirm.no-pending-commands").arguments(PLUGIN_PREFIX)
                )
            }
        )

        commandConfirmationManager.registerConfirmationProcessor(paperCommandManager)
        return commandConfirmationManager
    }

    private fun buildHelpSystem(): MinecraftHelp<CommandSender> {
        val help = MinecraftHelp.createNative("/pandorascluster help", paperCommandManager)
        help.helpColors = MinecraftHelp.HelpColors.of(
            NamedTextColor.DARK_GREEN,
            NamedTextColor.GREEN,
            NamedTextColor.BLUE,
            NamedTextColor.DARK_BLUE,
            NamedTextColor.AQUA
        )
        return help
    }

    private fun registerCommands() {

        annotationParser.parse(LandPlayerParser(plugin.api))
        annotationParser.parse(LandFlagParser(plugin.api))

        annotationParser.parse(LandTeleportCommands(plugin.api))
        annotationParser.parse(LandToggleBorderCommand(plugin.api, plugin))
        annotationParser.parse(SetHomeCommand(plugin.api))
        annotationParser.parse(SetFlagCommand(plugin.api))
        annotationParser.parse(ClaimCommand(plugin.api, plugin))
        annotationParser.parse(UnclaimCommand(plugin.api))
        annotationParser.parse(SetOwnerCommand(plugin.api))
        annotationParser.parse(SetRoleCommand(plugin.api))
        annotationParser.parse(LandInfoCommand(plugin.api))

        val builder = paperCommandManager.commandBuilder("land")
        paperCommandManager.command(
            builder.literal("confirm").meta(CommandMeta.DESCRIPTION, "Confirm")
                .handler(confirmationManager.createConfirmationExecutionHandler())
        )
    }
}